proposed/2021/LicenseInspection.md

```diff
@@ -0,0 +1,133 @@
+
+To avoid accidentally consuming an unsupported license we need a way to view the list and ideally fail builds when we use one of these licenses.
+
+### Functional explanation
+
+#### Local scenario
+
+You have just added a new NuGet dependency into the project and want to ensure it doesn't bring in any unsupported licenses.
+
+From the command line you run the license command:
+
+```bash
+$> dotnet nuget license
+```
+
+This outputs a human readable view of the license information:
+
+```
+Project dotnet-delice
+License Expression: MIT
+├── There are 10 occurances of MIT
+├─┬ Conformance:
+│ ├── Is OSI Approved: true
+│ ├── Is FSF Free/Libre: true
+│ └── Included deprecated IDs: false
+└─┬ Packages:
+  ├── FSharp.Core
+  ├── Microsoft.NETCore.App
+  ├── Microsoft.NETCore.DotNetAppHost
+  ├── Microsoft.NETCore.DotNetHostPolicy
+  ├── Microsoft.NETCore.DotNetHostResolver
+  ├── Microsoft.NETCore.Platforms
+  ├── Microsoft.NETCore.Targets
+  ├── NETStandard.Library
+  ├── Newtonsoft.Json
+  └── System.ComponentModel.Annotations
+```
+
+Satisfied with the licenses, you continue on with the task at hand.
+
+#### Blocking builds
+
+It's time to update your build pipeline to avoid releasing anything changes that introduce unsupported licenses. A new step is added to the pipeline YAML:
+
+```yml
+- name: Validate Licenses
+  runs: |
+    LICENSES=$(dotnet nuget license --json | jq '<query for allowed licenses>')
+    if ($LICENSES) then
+        exit 1
+    fi
+```
+
+This (pseduocode version) will dump the licenses to a JSON structure that can be parsed with a tool such as [jq](https://stedolan.github.io/jq/) and inspected for invalid licenses. If any are found an error code is returned, stopping the build from completing successfully.
+
+### Technical explanation
+
+The first technical challenge for this is the inconsistent nature of which licenses are provided by NuGet packages. While the [`licenseUrl` field was deprecated](https://github.com/NuGet/Announcements/issues/32), some projects haven't adopted the new format (or older packages that predate the deprecation are in use), making it difficult to determine what the license of a project is.
+
+The next challenge is how to detect licenses from license files. The ideal approach would be to mirror GitHub's approach, which uses [Licensee](https://licensee.github.io/licensee/) [for detection](https://help.github.com/en/articles/licensing-a-repository#detecting-a-license) (but naturally a dotnet implementation). Essentially, this uses [Sørensen–Dice coefficient](https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient) with a threshold for what is the acceptable level of comparison between the package's license file and license template.
+
+The technical workflow for license detection would follow:
+
+- If a `license` field is in the nuspec, check if it's a SPDX ID, if so, return. If it's a file, use Sørensen–Dice coefficient to compare it to a known list
```

Comment:
I am curious on why do you chose Dice coefficient for text similarity. Have you considered other text similarity techniques? For example
- [Jaccard index](https://en.wikipedia.org/wiki/Jaccard_index)
- [Cosine similarity](https://en.wikipedia.org/wiki/Cosine_similarity)

(I did a [homework on text similarity](https://github.com/dominoFire/deiis-114297-hw4/blob/master/src/main/resources/doc/hw4-report.pdf) during my college years)

Are we expecting large texts so that [min-hashing](https://en.wikipedia.org/wiki/MinHash) is worth doing? 

Also, how license text will be processed? Will stop words be removed?

Reply:
The reason that similarity algorithm is specified is because `dotnet-delice` (which inspired this proposal) is a .NET port of the JavaScript [`delice`](https://github.com/cutenode/delice) tool, and I wanted to have the same level of similarity applied so you could use both tools together and get comparable results.

License text should be processed as provided by the package, removal of stop words or anything else would mean that you're not accurately processing the license as shipped by the package and may give incorrect results.

The same goes for the license as provided by SPDX, there should be no modification of it to avoid incorrect results.